File: arc/cpp/runtime/factory.cpp
// Copyright 2025 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in
// the file licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business
// Source License, use of this software will be governed by the Apache License,
// Version 2.0, included in the file licenses/APL.txt.

#include "arc/cpp/runtime/factory.h"

namespace arc {

void MultiFactory::add(std::unique_ptr<NodeFactory> factory) {
    factories_.push_back(std::move(factory));
}

std::pair<std::unique_ptr<Node>, xerrors::Error>
MultiFactory::create(const NodeFactoryConfig& cfg) {
    for (auto& factory : factories_) {
        auto [node, err] = factory->create(cfg);

        if (!err) {
            // Success! This factory created the node
            return {std::move(node), xerrors::NIL};
        }

        if (err.type != "NOT_FOUND") {
            // Real error (not just "can't handle this type")
            // Add context and return immediately
            return {nullptr, xerrors::Error(
                err,
                err.data + " (while creating node '" + cfg.ir_node.key +
                "' of type '" + cfg.ir_node.type + "')")};
        }

        // NOT_FOUND: try next factory in chain
    }

    // No factory could handle this node type
    return {nullptr, xerrors::Error(
        "NOT_FOUND",
        "No factory registered for node type '" + cfg.ir_node.type +
        "' (node: " + cfg.ir_node.key + ")")};
}

}  // namespace arc

File: arc/go/runtime/node/factory.go
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

package node

import (
	"context"

	"github.com/synnaxlabs/alamos"
	"github.com/synnaxlabs/arc/ir"
	"github.com/synnaxlabs/arc/program"
	"github.com/synnaxlabs/x/errors"
	"github.com/synnaxlabs/x/query"
)

// Config provides dependencies and context for creating node instances.
type Config struct {
	alamos.Instrumentation
	// Node is the IR definition for this node.
	Node ir.Node
	// State provides access to input/output data and channel I/O.
	State *State
	// Program contains the arc program for accessing global state and functions.
	Program program.Program
}

// Factory creates node instances from IR definitions.
// Implementations check the node type and return query.NotFound if they cannot
// handle the given type.
type Factory interface {
	// Create constructs a node from the given configuration.
	// Returns query.NotFound if this factory cannot handle cfg.Node.Type.
	Create(ctx context.Context, cfg Config) (Node, error)
}

// CompoundFactory tries each factory in order until one succeeds. A factory that
// returns query.ErrNotFound is skipped; any other error is returned immediately.
type CompoundFactory []Factory

func (f CompoundFactory) Create(ctx context.Context, cfg Config) (Node, error) {
	for _, factory := range f {
		n, err := factory.Create(ctx, cfg)
		if err == nil {
			return n, nil
		}
		if errors.Is(err, query.ErrNotFound) {
			continue
		}
		return nil, err
	}
	return nil, query.ErrNotFound
}
